### PR TITLE
8328938: C2 SuperWord: disable vectorization for large stride and scale

### PR DESCRIPTION
**Pull request by eme64**
**Problem**
In [JDK-8310190](https://bugs.openjdk.org/browse/JDK-8310190) / https://git.openjdk.org/jdk/pull/14785 I fixed the alignment with `AlignVector`. For that, I had to compute `abs(scale)` and `abs(stride)`, as well as `scale * stride`. The issue is that all of these values can overflow the int range (e.g. `abs(min_int) = min_int`).

We hit asserts like:

`#  assert(is_power_of_2(value)) failed: value must be a power of 2: 0xffffffff80000000`
Happens because we take `abs(min_int)`, which is `min_int = 0x80000000`, and assuming this was a positive (unsigned) number is a power of 2 `2^31`. We then expand it to `long`, get `0xffffffff80000000`, which is not a power of 2 anymore. This violates the implicit assumptions, and we hit the assert.

`#  assert(q >= 1) failed: modulo value must be large enough`
We have `scale = 2^30` and `stride = 4 = 2^2`. For the alignment calculation we compute `scale * stride = 2^32`, which overflows the int range and becomes zero.

Before [JDK-8310190](https://bugs.openjdk.org/browse/JDK-8310190) we could get similar issues with the (old) code in `SuperWord::ref_is_alignable`, if `AlignVector` is enabled:

```
int span = preloop_stride * p.scale_in_bytes();
...
if (vw % span == 0) {
```

if `span == 0` because of overflow, then the `idiv` from the modulo gets a division by zero -> `SIGFPE`.

But it seems the bug is possibly a regression from JDK20 b2 [JDK-8286197](https://bugs.openjdk.org/browse/JDK-8286197). Here we enabled certaint Unsafe memory access address patterns, and it is such patterns that the reproducer requires.

**Solution**
I could either patch up all the code that works with `scale` and `stride`, and make sure no overflows ever happen. But that is quite involved and error prone.

I now just disable vectorization for large `scale` and `stride`. This should not have any performance impact, because such large `scale` and `stride` would lead to highly inefficient memory accesses, since they are spaced very far apart.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8328938](https://bugs.openjdk.org/browse/JDK-8328938): C2 SuperWord: disable vectorization for large stride and scale (**Bug** - P3)

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18485/head:pull/18485` \
`$ git checkout pull/18485`

Update a local copy of the PR: \
`$ git checkout pull/18485` \
`$ git pull https://git.openjdk.org/jdk.git pull/18485/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18485`

View PR using the GUI difftool: \
`$ git pr show -t 18485`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18485.diff">https://git.openjdk.org/jdk/pull/18485.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18485#issuecomment-2022046339)

### CODE REVIEW

File: src/hotspot/share/opto/vectorization.cpp [hunk 1]
```diff
@@ -393,6 +393,25 @@ VPointer::VPointer(const MemNode* mem, const VLoop& vloop,
   NOT_PRODUCT(if(_tracer._is_trace_alignment) _tracer.restore_depth();)
   NOT_PRODUCT(_tracer.ctor_6(mem);)
 
+  // In the pointer analysis, and especially the AlignVector analysis we assume that
```
Comment (chhagedorn) [inline src/hotspot/share/opto/vectorization.cpp:396]:
```suggestion
  // In the pointer analysis, and especially the AlignVector, analysis we assume that
```

File: src/hotspot/share/opto/vectorization.cpp [hunk 2]
```diff
@@ -393,6 +393,25 @@ VPointer::VPointer(const MemNode* mem, const VLoop& vloop,
...
+  // stride and scale are not too large. For example, we multiply "scale * stride",
+  // and assume that this does not overflow the int range. We also take "abs(scale)"
+  // and "abs(stride)", which would overflow for min_int = -(2^31). Still, we want
+  // to at least allow small and moderately large stride and scale. Therefore, we
+  // allow values up to 2^30, which is only a factor 2 smaller than the max/min int
+  // values. Normal performance relevant code will have much lower. And the restriction
```
Comment (chhagedorn) [inline src/hotspot/share/opto/vectorization.cpp:402]:
```suggestion
  // allow values up to 2^30, which is only a factor 2 smaller than the max/min int.
  // Normal performance relevant code will have much lower values. And the restriction
```

**Review by chhagedorn — APPROVED** [review]
That's a reasonable, safe and easy solution.

**Review by chhagedorn — APPROVED** [review]

File: src/hotspot/share/opto/vectorization.cpp [hunk 3]
```diff
@@ -393,6 +393,25 @@ VPointer::VPointer(const MemNode* mem, const VLoop& vloop,
...
+  // Normal performance relevant code will have much lower values. And the restriction
+  // allows us to keep the rest of the autovectorization code much simpler, since we
+  // do not have to deal with overflows.
+  jlong long_scale  = _scale;
+  jlong long_stride = _vloop.iv_stride();
+  jlong max_val = 1 << 30;
+  if (abs(long_scale) >= max_val ||
+      abs(long_stride) >= max_val ||
+      abs(long_scale * long_stride) >= max_val) {
+    assert(!valid(), "adr stride*scale is too large");
```
[the remark this answers is not in the document]
Comment (eme64) [reply src/hotspot/share/opto/vectorization.cpp:411]: If you look a few lines up, you can see that all other "bailouts" also check that the VPointer is invalid. I am simply matching the surrounding code. And it also makes it explicit, that the VPointer will be invalid, which is what I want.

[the remark this answers is not in the document]
Comment (vnkozlov) [reply src/hotspot/share/opto/vectorization.cpp:411]: okay

**Review by vnkozlov — APPROVED** [review]

**Comment by eme64** [comment]
@chhagedorn @vnkozlov thanks for the reviews!
/integrate

**Pull request closed without merging**